### PR TITLE
Allows read for certificates and their keys

### PR DESCRIPTION
**Pull request by mettke**
This PR addresses #704 and requires https://github.com/SergioBenitez/hyper-sync-rustls/pull/5

With the PR it is possible to use custom Files, Arrays, String and pretty much everything else that implements `Read` or can be used in conjunction with `Cursor` for SSL certificates.

### Example for custom file

```rust
let mut config = Config::active().unwrap();
let certfile = fs::File::open("certs/server.crt").unwrap();
let keyfile = fs::File::open("certs/server.key").unwrap();
config.set_tls_from_read(certfile, keyfile);
```

### Example for byte vector

```rust
let mut config = Config::active().unwrap();
let certfile_bytes = Cursor::new(include_bytes!("certs/server.crt").to_vec());
let keyfile_bytes = Cursor::new(include_bytes!("certs/server.key").to_vec());
config.set_tls_from_read(certfile, keyfile);
```

### Example for string

```rust
let mut config = Config::active().unwrap();
let certfile = fs::File::open("certs/server.crt").unwrap();
let keyfile = fs::File::open("certs/server.key").unwrap();
let mut certfile_str = String::new();
let mut keyfile_str = String::new();
certfile.read_to_string(&mut certfile_str);
keyfile.read_to_string(&mut keyfile_str);
config.set_tls_from_read(Cursor::new(certfile_str), Cursor::new(keyfile_str));
```

### CODE REVIEW

**Comment by mettke** [comment]
Build failure due to the missing PR in hyper-sync-rustls

**Comment by SergioBenitez** [comment]
@mettke Please point `hyper-sync-rustls` to your repository with the prerequisite changes for review.

**Comment by mettke** [comment]
@SergioBenitez Very well - Done

File: core/lib/src/config/config.rs [hunk 1]
```diff
@@ -577,6 +578,56 @@ impl Config {
...
+    /// # fn config_test() -> Result<(), ConfigError> {
+    /// let mut config = Config::development()?;
+    /// let certfile_bytes = Cursor::new([10, 10, 10, 10].to_vec());
+    /// let keyfile_bytes = Cursor::new([10, 10, 10, 10].to_vec());
+    /// config.set_tls_from_read(certfile_bytes, keyfile_bytes)?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    #[cfg(feature = "tls")]
+    pub fn set_tls_from_read<R: Read + Seek>(&mut self, certs_read: R, key_read: R) -> Result<()> {
```
Comment (jebrosen) [inline core/lib/src/config/config.rs]: (bikeshed) IMO names like `certs_data` and `key_data` are more natural `certs_read`.

File: core/lib/src/config/config.rs [hunk 2]
```diff
@@ -577,6 +578,56 @@ impl Config {
...
+    ///
+    /// Certificates are read from `certs_read`. The certificate chain must be
+    /// in X.509 PEM format. The private key is read from `key_read`. The
+    /// private key must be an RSA key in either PKCS#1 or PKCS#8 PEM format.
+    ///
+    /// # Errors
+    ///
+    /// If reading either the certificates or private key fails, an error of
+    /// variant `Io` is returned. If either the certificates or private key
+    /// files are malformed or cannot be parsed, an error of `BadType` is
```
Comment (jebrosen) [inline core/lib/src/config/config.rs]: 'files' -> 'data'

File: core/lib/src/config/config.rs [hunk 3]
```diff
@@ -577,6 +578,56 @@ impl Config {
...
+    /// # Example
+    ///
+    /// ```rust
+    /// use rocket::config::Config;
+    /// use std::io::Cursor;
+    ///
+    /// # use rocket::config::ConfigError;
+    /// # fn config_test() -> Result<(), ConfigError> {
+    /// let mut config = Config::development()?;
+    /// let certfile_bytes = Cursor::new([10, 10, 10, 10].to_vec());
```
Comment (jebrosen) [inline core/lib/src/config/config.rs]: Does this test even succeed? I can't imagine that data is valid.

I have misgivings about this example; embedding certs inside the source code is not something I would want to encourage.

File: core/lib/src/config/config.rs [hunk 4]
```diff
@@ -577,6 +578,56 @@ impl Config {
...
+    /// # Ok(())
+    /// # }
+    /// ```
+    #[cfg(feature = "tls")]
+    pub fn set_tls_from_read<R: Read + Seek>(&mut self, certs_read: R, key_read: R) -> Result<()> {
+        use http::tls::util::{self, Error};
+
+        let pem_err = "malformed PEM file";
+
+        // Load the certificates.
```
Comment (jebrosen) [inline core/lib/src/config/config.rs]: This all duplicates the existing non-`Read` variant pretty closely, which isn't the greatest.

**Comment by SergioBenitez** [comment]
`Config` is about to change drastically with #852. Instead of moving forward with this PR, we should ensure that the revamped configuration for #852 allows for something like this easily.

**Comment by SergioBenitez** [comment]
Let's take care of this when we merge a solution for #852.

**Pull request closed without merging**